osquery/carver/carver.cpp

```diff
@@ -0,0 +1,286 @@
+    return;
+  }
+
+  tree.put(key, value);
+
+  std::ostringstream os;
+  pt::write_json(os, tree, false);
+  setDatabaseValue(kQueries, kCarverDBPrefix + guid, os.str());
+}
+
+Carver::Carver(const std::set<std::string>& paths, const std::string& guid) {
+  for (const auto& p : paths) {
+    carvePaths_.insert(fs::path(p));
+  }
+
+  // Construct the uri we post our data back to:
+  startUri_ = TLSRequestHelper::makeURI(FLAGS_carver_start_endpoint);
+  contUri_ = TLSRequestHelper::makeURI(FLAGS_carver_continue_endpoint);
+
+  // Generate a unique identifier for this carve
+  carveGuid_ = guid;
+
+  // TODO: Adding in a manifest file of all carved files might be nice.
+  carveDir_ =
+      fs::temp_directory_path() / fs::path(kCarvePathPrefix + carveGuid_);
+  auto ret = fs::create_directory(carveDir_);
+  if (!ret) {
+    LOG(ERROR) << "Unable to create carve file store";
+  }
+
+  // Store the path to our archive for later exfiltration
+  archivePath_ = carveDir_ / fs::path(kCarveNamePrefix + carveGuid_ + ".tgz");
+
+  // Update the DB to reflect that the carve is pending.
+  updateCarveValue(carveGuid_, "status", "PENDING");
+};
+
+Carver::~Carver() {
+  fs::remove_all(carveDir_);
+}
+
+void Carver::start() {
+  for (const auto& p : carvePaths_) {
+    if (!fs::exists(p)) {
+      LOG(WARNING) << "File does not exist on disk: " << p;
+    } else {
+      Status s = carve(p);
+      if (!s.ok()) {
+        LOG(WARNING) << "Failed to carve file: " << p;
+      }
+    }
+  }
+
+  std::set<fs::path> carvedFiles;
+  for (const auto& p : platformGlob((carveDir_ / "*").string())) {
+    carvedFiles.insert(fs::path(p));
+  }
+
+  auto s = compress(carvedFiles);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to create carve archive: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+
+  s = exfil(archivePath_);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to post carve: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+};
+
+Status Carver::carve(const boost::filesystem::path& path) {
+  PlatformFile src(path.string(), PF_OPEN_EXISTING | PF_READ);
+  PlatformFile dst((carveDir_ / path.leaf()).string(),
+                   PF_CREATE_NEW | PF_WRITE);
+
+  if (!dst.isValid()) {
+    return Status(1, "Destination tmp FS is not valid.");
+  }
+
+  auto blkCount =
+      ceil(static_cast<double>(src.size()) / static_cast<double>(kBuffSize));
+
+  for (size_t i = 0; i < blkCount; i++) {
+    std::vector<char> inBuff(kBuffSize, 0);
+    auto bytesRead = src.read(inBuff.data(), kBuffSize);
+    auto bytesWritten = dst.write(inBuff.data(), bytesRead);
+    if (bytesWritten < 0) {
+      return Status(1, "Error writing bytes to tmp fs");
+    }
+  }
+
+  return Status(0, "Ok");
+};
+
+Status Carver::compress(const std::set<boost::filesystem::path>& paths) {
+  auto arch = archive_write_new();
+  archive_write_set_format_zip(arch);
```

Comment:
Absolutely, however currently this adds more or less nothing but the zip headers and packaging. There's currently no actual deflation of data, and I'm not sure what's wrong with how I'm using `libarchive`. At this point this line could probably even be removed.